packages/by-name/kata/kata-runtime/0013-genpolicy-allow-image_guest_pull.patch

```diff
@@ -0,0 +1,1471 @@
+     p_namespace == i_namespace
+ 
+     allow_by_container_types(p_oci, i_oci, s_name, p_namespace)
+-    allow_by_bundle_or_sandbox_id(p_oci, i_oci, p_storages, i_storages)
++    allow_by_bundle_or_sandbox_id(p_container, i_oci, i_storages)
+     allow_process(p_oci, i_oci, s_name)
+ 
+     print("allow_by_sandbox_name: true")
+@@ -504,11 +505,12 @@ allow_linux_sysctl(p_linux, i_linux) {
+ 
+ # Check the consistency of the input "io.katacontainers.pkg.oci.bundle_path"
+ # and io.kubernetes.cri.sandbox-id" values with other fields.
+-allow_by_bundle_or_sandbox_id(p_oci, i_oci, p_storages, i_storages) {
++allow_by_bundle_or_sandbox_id(p_container, i_oci, i_storages) {
+     print("allow_by_bundle_or_sandbox_id: start")
+ 
+     key := "io.kubernetes.cri.sandbox-id"
+ 
++    p_oci := p_container.OCI
+     p_regex := p_oci.Annotations[key]
+     sandbox_id := i_oci.Annotations[key]
+ 
+@@ -528,8 +530,7 @@ allow_by_bundle_or_sandbox_id(p_oci, i_oci, p_storages, i_storages) {
+         allow_mount(p_oci, i_mount, bundle_id, sandbox_id)
+     }
+ 
+-    # TODO: enable allow_storages() after fixing https://github.com/kata-containers/kata-containers/issues/8833
+-    # allow_storages(p_storages, i_storages, bundle_id, sandbox_id)
++    allow_storages(p_container, i_storages, bundle_id, sandbox_id)
+ 
+     print("allow_by_bundle_or_sandbox_id: true")
+ }
+@@ -824,30 +825,75 @@ mount_source_allows(p_mount, i_mount, bundle_id, sandbox_id) {
+ ######################################################################
+ # Create container Storages
+ 
+-allow_storages(p_storages, i_storages, bundle_id, sandbox_id) {
++# Allow image_guest_pull storage
++allow_storages(p_container, i_storages, bundle_id, sandbox_id) {
++    print("allow_storages 1: start")
++    i_count := count(i_storages)
++    print("allow_storages 1: i_count =", i_count)
++    i_count == 1
++
++    i_storage := i_storages[0]
++    i_storage.driver == "image_guest_pull"
++    i_storage.fstype == "overlay"
++
++    p_mount_point := concat("/", [policy_data.common.cpath, bundle_id, "rootfs"])
++    print("allow_storages 1: i_storage.mount_point =", i_storage.mount_point, "p_mount_point =", p_mount_point)
++    i_storage.mount_point == p_mount_point
++    
++    print("allow_storages 1: p_container.image =", p_container.image, "i_storage.source =", i_storage.source)
++
++    count(i_storage.driver_options) == 1
++    i_driver_option := i_storage.driver_options[0]
++
++    i_driver_options_json := trim_prefix(i_driver_option, "image_guest_pull=")
++    print("allow_storages 1: i_driver_options_json =", i_driver_options_json)
++    i_driver_options := json.unmarshal(i_driver_options_json)
++
++    allow_guest_pull(p_container.image, i_storage.source, i_driver_options)
++
++    print("allow_storages 1: true")
++}
++
++allow_guest_pull(p_container_image, i_storage_source, i_driver_options) {
++    # pause container
++    i_storage_source == "pause"
```

Comment:
FWIW the kata-agent uses different logic for detecting pause image references:
https://github.com/kata-containers/kata-containers/blob/13517cf9c1680740c9429b85bbbc70bc1d2ed8f5/src/agent/src/image.rs#L137-L145

Reply:
Let me add a check for `container-type == "sandbox"` here for defense-in-depth (i.e., pause image is not accidentally used for regular container). I'd still want to keep the other two checks, though, to prevent regressions if something else should start interpreting these fields in the future.

Reply:
Check added